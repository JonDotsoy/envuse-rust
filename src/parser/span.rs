use serde::{Serialize, Deserialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}

impl Span {
    pub fn substring<T: ToString>(&self, payload: T) -> String {
        payload.to_string()[self.start..self.end].to_string()
    }
}
