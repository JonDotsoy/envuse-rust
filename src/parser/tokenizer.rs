use std::{ops::Add, usize, vec};

#[derive(Debug)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}

#[derive(Debug)]
pub struct Token {
    pub kind: String,
    pub raw: String,
    pub span: Span,
}

impl Token {
    fn new<A: ToString>(kind: A, cursor: &Cursor, span: Span) -> Self {
        Self {
            kind: kind.to_string(),
            raw: cursor.get_by_span(&span),
            span,
        }
    }
}

#[derive(Debug)]
struct Cursor {
    payload: String,
    index: usize,
}

impl Cursor {
    fn new<A>(index: usize, payload: A) -> Self
    where
        A: ToString,
    {
        Self {
            payload: payload.to_string(),
            index,
        }
    }

    fn has_next<'a>(&'a self) -> bool {
        (self.payload.len() > (self.index + 1)).clone()
    }

    fn has_current<'a>(&'a self) -> bool {
        self.payload.len() > self.index
    }

    fn current_matches_char(&self, val: char) -> bool {
        if let Some(c) = self.current_char() {
            c == val
        } else {
            false
        }
    }

    fn next_matches_char(&self, val: char) -> bool {
        if let Some(c) = self.next_char() {
            c == val
        } else {
            false
        }
    }

    fn next_matches_range_char(&self, range: std::ops::RangeInclusive<char>) -> bool {
        matches!(self.next_char(), range)
    }

    fn next_char(&self) -> Option<char> {
        self.payload.chars().nth(self.index + 1)
    }

    fn current_char(&self) -> Option<char> {
        self.payload.chars().nth(self.index)
    }

    fn forward(&mut self, positions: usize) {
        self.index = self.index + positions;
    }

    fn current_char_expected(&self, val: char) {
        if let Some(c) = self.current_char() {
            if c != val {
                todo!("Require error expected the char {}", val);
            }
        } else {
            todo!("Require error: cannot found current char");
        }
    }

    fn get_by_span(&self, span: &Span) -> String {
        unsafe { self.payload.get_unchecked(span.start..span.end).to_string() }
    }
}

pub struct Tokenizer {}

impl Tokenizer {
    pub fn parse<A>(payload: A) -> Vec<Token>
    where
        A: ToString,
    {
        let ref mut cursor = Cursor::new(0, payload);
        return Self::parse_by_cursor(cursor);
    }

    fn parse_by_cursor(cursor: &mut Cursor) -> Vec<Token> {
        let mut tokens: Vec<Token> = vec![];

        while cursor.has_current() {
            if cursor.current_matches_char('#') {
                tokens.extend(Self::parse_comment(cursor));
                continue;
            }

            if cursor.current_matches_char('\n') {
                tokens.extend(Self::parse_new_line(cursor));
                continue;
            }

            dbg!(&cursor);
            dbg!(&cursor.current_char());
            dbg!(&tokens);
            todo!();
        }

        tokens
    }

    fn parse_comment(cursor: &mut Cursor) -> Vec<Token> {
        cursor.current_char_expected('#');
        let span_start = cursor.index;

        while cursor.has_current() {
            cursor.forward(1);
            if cursor.current_matches_char('\n') {
                break;
            }
        }

        let span = Span {
            start: span_start,
            end: cursor.index,
        };
        let comment_token = Token::new("comment", &cursor, span);
        vec![comment_token]
    }

    fn parse_new_line(cursor: &mut Cursor) -> Vec<Token> {
        cursor.current_char_expected('\n');
        let span_start = cursor.index;
        cursor.forward(1);
        let span = Span {
            start: span_start,
            end: cursor.index,
        };
        let newline_token = Token::new("newline", &cursor, span);
        vec![newline_token]
    }
}
