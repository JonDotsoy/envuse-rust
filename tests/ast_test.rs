#[cfg(test)]
mod ast_test {
    use envuse_parser::parser::span::Span;
    use envuse_parser::parser::tokenizer::Tokenizer;

    use envuse_parser::{parser::ast::AST, utils::display_syntax::DisplaySyntax};
    use insta::{assert_debug_snapshot, assert_snapshot, assert_yaml_snapshot};

    #[test]
    fn parse_executable() {
        let payload = "#!/bin/sh";
        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }

    #[test]
    fn parse_comment() {
        let payload = "# I'm comment";
        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }

    #[test]
    fn parse_comment_block() {
        let payload = "# I'm comment\n# Second line comment";
        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }

    #[test]
    fn parse_comment_multiple_block() {
        let payload = "# I'm comment\n\n# Second line comment";
        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }

    #[test]
    fn parse_comment_multiple_lines() {
        let payload = "# I'm comment\n    # Second line comment";
        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }

    #[test]
    fn parse_comment_multiple_block_2() {
        let payload = "# I'm comment\n\n     # Second line comment";
        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }

    #[test]
    fn parse_variables_name() {
        let payload = "
            foo
        ";
        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }

    #[test]
    fn parse_variables_with_comment() {
        let payload = "
            # comment
            foo
        ";
        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }

    #[test]
    fn parse_variables_with_type() {
        let payload = "
            foo: string
        ";
        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }

    #[test]
    fn parse_variables_as_nullable() {
        let payload = "
            FOO: String<Foo Biz> ?
        ";
        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }

    #[test]
    fn parse_variable_with_default_value() {
        let payload = r#"foo = "abc""#;
        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }

    #[test]
    fn parse_variable_with_default_value_number() {
        let payload = r#"port: number = 3000"#;
        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }

    #[test]
    fn parse_variable() {
        let payload = r#"
            # comment
            foo: string = "abc"
        "#;
        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }

    #[test]
    fn parse_sample_document() {
        let payload = r#"
            ## Document comment
            ## second line

            # comment
            foo: string = "abc"
            bax: number = "123"
            port: number = 3_000
        "#;
        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }

    #[test]
    fn parse_unexpected_token() {
        let payload = r#"

            _$3

        "#;

        let err = Tokenizer::parse(payload).err();
        assert_debug_snapshot!(err);
    }

    // fn s<T: ToString, F: FnOnce() -> Result<(), SyntaxError>>(payload: &T, f: F) {
    //     match f() {
    //         Ok(_) => {}
    //         Err(err) => {
    //             // dbg!(err.debug_payload(&*payload))
    //         }
    //     }
    // }

    #[test]
    fn should_decorate_syntax_error() {
        let payload = "#!/bin/envuse\nFOO: String<Max=500>\nBIZ: String<Max=500 Min=2>\n";

        assert_snapshot!(
            DisplaySyntax::new("SyntaxError: fail", Span { start: 14, end: 17 })
                .debug_payload(&payload)
        );
    }

    #[test]
    fn should_decorate_syntax_error_2() {
        let payload =
            "#!/bin/envuse\n# Inline comment\nFOO: String<Max=500>\nBIZ: String<Max=500 Min=2>\nPORT: Number\n";

        let err = DisplaySyntax::new("SyntaxError: fail", Span { start: 36, end: 83 });

        // dbg!(err.span.substring(&payload));

        assert_snapshot!(err.debug_payload(&payload));
    }

    #[test]
    fn parse_sample_document_with_type_configurable() {
        let payload = r#"
            FOO : String<Max=500>
            BIZ : String<Max=500 Min=2>
        "#;

        let tokens = Tokenizer::parse(payload).unwrap();
        assert_yaml_snapshot!(AST::parse(tokens).unwrap());
        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }

    #[test]
    fn parse_sample_document_with_type_configurable_binary_option() {
        let payload = r#"
            FOO : String<Max=500 Sensitive>
            BIZ : String<Max=500 Min=2>
        "#;

        let tokens = Tokenizer::parse(payload).unwrap();
        assert_yaml_snapshot!(AST::parse(tokens).unwrap());
        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }

    #[test]
    fn parse_sample_document_with_type_configurable_multiline() {
        let payload = r#"
            FOO : String<
                    Max = 500
                    Min = 2
                  >
        "#;

        let tokens = Tokenizer::parse(payload).unwrap();
        assert_debug_snapshot!(AST::parse(tokens).unwrap());
    }
}
