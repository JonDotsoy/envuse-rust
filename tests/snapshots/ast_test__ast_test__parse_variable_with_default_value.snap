---
source: tests/ast_test.rs
expression: "AST::parse(tokens).unwrap()"
---
Document {
    span: Span {
        start: 0,
        end: 10,
    },
    executable: None,
    elements: [
        Variable {
            span: Span {
                start: 0,
                end: 10,
            },
            comment: None,
            name: "foo",
            variable_type: None,
            default_value: Some(
                DefaultValue {
                    span: Span {
                        start: 7,
                        end: 10,
                    },
                    value: "abc",
                },
            ),
        },
    ],
}
